core/src/main/java/org/jruby/internal/runtime/ThreadService.java

```diff
@@ -202,30 +197,34 @@ public void initMainThread() {
      * collected.
      */
     public final ThreadContext getCurrentContext() {
-        SoftReference<ThreadContext> ref;
-        ThreadContext context = null;
-
-        while (context == null) {
-            // loop until a context is available, to clean up softrefs that might have been collected
-            if ((ref = localContext.get()) == null) {
-                ref = adoptCurrentThread();
-                context = ref.get();
-            } else {
-                context = ref.get();
-            }
-
-            // context is null, wipe out the SoftReference (this could be done with a reference queue)
-            if (context == null) {
-                localContext.set(null);
-            }
+        SoftReference<ThreadContext> ref = localContext.get();
+        if (ref == null) {
+            adoptCurrentThread(); // registerNewThread will localContext.set(...)
+            ref = localContext.get();
+        }
+        ThreadContext context;
+        if ((context = ref.get()) != null) {
+            return context;
         }
 
-        return context;
+        // context is null, wipe out the SoftReference (this could be done with a reference queue)
+        localContext.set(null);
+
+        // loop until a context is available, to clean up soft-refs that might have been collected
+        return getCurrentContext();
```

Comment:
Isn't this recursing? Unlikely to be a problem but the comment claims it's a loop.

Reply:
there used to be one - so the comment was kept, except for the loop word its true :)